### PR TITLE
x/gamm: Reduce swap fee for Osmo-routed trades (#2406)

### PR DESCRIPTION
**Pull request by hleb-albau**
Closes: #2406

## What is the purpose of the change

This pull request reduce swap fee for a half during calculation for each pool for routes with length of 2 and osmo intermediate, ex `ATOM -> OSMO -> USDC`. 

In general it is re-implementation of https://github.com/osmosis-labs/osmosis/pull/1069 with addition tests added.

## Brief Changelog

  - x/gamm Change `SwapExactAmountIn` and `SwapExactAmountOut`to take fee multiplier as addition input param.
  - Update multihop swaps test, to insure that half fee applies only for osmo routed multihop swaps.

## Testing and Verifying

This change added tests and can be verified as follows:

  - *Added cases for existing multihop unit tests.*

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes 
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`?  yes
  - How is the feature or change documented? specification (`x/<module>/spec/`)

### CODE REVIEW

File: x/gamm/README.md
Comment (stackman27) [inline x/gamm/README.md:91]: Thoughts on adding an example here?

Also, does it matter if we half the `swapFee` at the end (totalfee / 2) vs halving it per pool swap with usmo?

File: CHANGELOG.md
Comment (stackman27) [inline CHANGELOG.md:61]: Duplicate changelog

```suggestion
```

File: x/gamm/keeper/multihop_test.go
Comment (stackman27) [inline x/gamm/keeper/multihop_test.go:289]: Does it matter if `reducedFee` is applied or not if the swap fee is 0?

Comment (stackman27) [reply x/gamm/keeper/multihop_test.go:289]: Cause i think `suite.Require().True(tokenInAmount.GT(tokenInCalculatedAsSeparateSwapsWithoutFee.Amount))` doesnot have to be in the if statement

Comment (hleb-albau) [reply x/gamm/keeper/multihop_test.go:289]: if swap fee is 0, than fee will not reduced. 
this line is just one more defensive check, that we will not lost user assets due some miscode with fee reduction (ex: after fee reduction another user could get more assets, that raw swap without fee at all)

Comment (hleb-albau) [reply x/gamm/keeper/multihop_test.go:289]: we can change it to `GTE`

Comment (stackman27) [reply x/gamm/keeper/multihop_test.go:289]: > (ex: after fee reduction another user could get more assets, that raw swap without fee at all)

How is this possible? If we have fee reduction, there is still a `swapFee` so, how can a user get more assets than `tokenInAmount`?

File: x/gamm/README.md
Comment (hleb-albau) [reply x/gamm/README.md:91]: Yes, in fact `totalfee` will be reduced by  less than by 50%, depending on various pool states. On test introduced in that PR it will be around 25%, cuz we create a pool with low liquidity.  In general final percentage will grow up to 50% the higher liquidity is.
I just made it as in proposal description(see forum) and sunny first implementation. I think it more elegant and stable solution, rather than trying to make it half for all cases.

File: x/gamm/keeper/multihop_test.go
Comment (hleb-albau) [reply x/gamm/keeper/multihop_test.go:289]: > > (ex: after fee reduction another user could get more assets, that raw swap without fee at all)
> 
> How is this possible? If we have fee reduction, there is still a `swapFee` so, how can a user get more assets than `tokenInAmount`?

just remember a june case, when users was able to get 50% from pools :) By design it is not possible, but one math code mistake, that was not covered my simple unit test make it real.

Comment (stackman27) [reply x/gamm/keeper/multihop_test.go:289]: aah makes sense! You got me there 😁

File: x/gamm/README.md
Comment (hleb-albau) [reply x/gamm/README.md:94]: added example in re-base commit

**Comment by hleb-albau** [comment]
Also feel free to edit pr on your own, if it can safe your time for merge.

**Comment by hleb-albau** [comment]
@stackman27 also we should create issue for frontend as well. But i don't know where to do it.

**Comment by stackman27** [comment]
> @stackman27 also we should create issue for frontend as well. But i don't know where to do it.

You can report frontend issues here (https://github.com/osmosis-labs/osmosis-frontend/issues)

**Comment by hleb-albau** [comment]
> > @stackman27 also we should create issue for frontend as well. But i don't know where to do it.
> 
> You can report frontend issues here (https://github.com/osmosis-labs/osmosis-frontend/issues)

Done
https://github.com/osmosis-labs/osmosis-frontend/issues/745

File: x/gamm/README.md
Comment (mattverse) [inline x/gamm/README.md:93]: asking to learn: why is this so?

File: x/gamm/keeper/multihop.go
Comment (mattverse) [inline x/gamm/keeper/multihop.go:69]: should it not be (5,2)?

File: x/gamm/types/route.go
Comment (mattverse) [inline x/gamm/types/route.go:11]: Two questions here:
1. Why do we have len == 2 here? What if route has length of three which all includes Osmo in the routes?
2. Why are we always taking 0 index of routes and expect it to be basecoin unit here?

File: x/gamm/keeper/multihop_test.go
Comment (mattverse) [inline x/gamm/keeper/multihop_test.go:139]: Can we have test cases that tests with exact numbers on reduce fees?

**Review by mattverse — COMMENTED** [review]
Nice implementation on this! Left some initial comments and reviews!

File: x/gamm/types/route.go
Comment (hleb-albau) [reply x/gamm/types/route.go:11]: In general the idea that osmo will have a lot of pools X-OSMO
Then to Swap X->Y  all swaps will looks like X -> OSMO -> Y
So to detect such swaps we take the first hop and check that it has output OSMO(that mean that input for the second hop will be osmo too)

What if route has length of three which all includes Osmo in the routes 
- That possible only if we have second hot OSMO->OSMO, or i miss something?

File: x/gamm/keeper/multihop.go
Comment (hleb-albau) [reply x/gamm/keeper/multihop.go:69]: (5,2) means 0.05
if we want to half fee, we need to use 0.5 multiplier

File: x/gamm/keeper/swap.go
Comment (czarcas7ic) [inline x/gamm/keeper/swap.go:33]: Is there a specific reason you decided to have the swapFeeMultiplier as an input to SwapExactAmountIn/Out? It seems like having the logic directly in the SwapExactAmountIn/Out function itself would be potentially safer/cleaner (I could also just totally be missing something!). Lmk what you think

File: x/gamm/keeper/multihop_test.go
Comment (czarcas7ic) [reply x/gamm/keeper/multihop_test.go:139]: Agreed, I don't think GT LT is sufficient for this

File: x/gamm/keeper/multihop_test.go
Comment (czarcas7ic) [inline x/gamm/keeper/multihop_test.go:292]: Same here (changing to specific values rather than GT LT)

File: x/gamm/keeper/swap.go
Comment (czarcas7ic) [inline x/gamm/keeper/swap.go:93]: I think we should protect against negative numbers here (ensuring it is not possible to set a negative swap fee since we are now essentially modifying the pools swap fee). What do you think?

**Review by czarcas7ic — COMMENTED** [review]
Well done on this! 

I would like to have a discussion on both having the swapFee reduction logic within SwapExactAmountIn/Out and protecting against negative swap fees. After that, IMO its good to go 😄

File: x/gamm/keeper/swap.go
Comment (hleb-albau) [reply x/gamm/keeper/swap.go:93]: I think this place is not where we should force and check for negative swaps fee. If we have here negative swap fee, than something is broken prior this code.

Or you mean, that `swapFeeMultiplier` can't be negative. Is so, that make sense, will add one more check here.

File: x/gamm/README.md
Comment (hleb-albau) [reply x/gamm/README.md:93]: ops, seems i just invent it. I try to debug and see real values -> it is always around 50%

Comment (hleb-albau) [reply x/gamm/README.md]: hm, i found where a got 25% - in test of TestBalancerPoolSimpleMultihopSwapExactAmountOut

So for out i always got around 25%. Can't understand now, fill try to figure it out.

Comment (hleb-albau) [reply x/gamm/README.md]: heh, more exact test gives you sometimes nice catches
found a bug in out fee reduction calculation, now it is the same as input - around 50%

File: x/gamm/keeper/swap.go
Comment (hleb-albau) [reply x/gamm/keeper/swap.go:33]: This functions performs only single swaps. That mean, them don't know context executions - whether it is multihop swaps or just single swap. But i feel something as you feel - that it is a bit strange to have that multiplier there. Wii try to create another version where multihop logic will use private keeper functions.

File: x/gamm/keeper/swap.go
Comment (hleb-albau) [reply x/gamm/keeper/swap.go]: Check out new version, where i remove this input param. So it is not actual anymore.

File: x/gamm/keeper/multihop_test.go
Comment (czarcas7ic) [inline x/gamm/keeper/multihop_test.go:122]: I think what you are doing here is choosing 0 as the poolIdShift so you utilize the actual poolId which then utilizes the real fee. You then use 2 as the poolIdShift to select a non existent pool which has no fee because it doesn't exist.

What do you think about instead exporting the internal swapExactAmountIn (this https://github.com/osmosis-labs/osmosis/blob/c1808cd052160f3d187693b042cfd09892fc4f11/x/gamm/keeper/swap.go#L39-L40) via the export_test.go file (you would need to export it as some other name since the exported SwapExactAmountIn already exists). Idk, I am just thinking of ways to make this more robust / less hacky. Lmk what you think!

If you disagree with this method, I think at the very least we need to explain the hacky method we are doing in the comments (I know you said "use +2 to calc using second pair (w/o swap fee)" but without context this is confusing and took me a bit of time to figure out what was actually going on here)

File: x/gamm/keeper/multihop_test.go
Comment (czarcas7ic) [inline x/gamm/keeper/multihop_test.go:136]: I think you meant "other stuff" here, can we either be more specific about what this other stuff is, or just remove this part at the end?

File: x/gamm/keeper/multihop_test.go
Comment (czarcas7ic) [inline x/gamm/keeper/multihop_test.go:142]: Would like to do some work here to make this slightly clearer, such as:

1) for diffB, subtract in the same order as we did for diffA (then add the differences rather than subtract)
2) add comments explaining what exactly diffA and diffB represent

File: x/gamm/keeper/multihop_test.go
Comment (czarcas7ic) [inline x/gamm/keeper/multihop_test.go:68]:
```suggestion
			// Prepare pools
```

File: x/gamm/keeper/multihop_test.go
Comment (czarcas7ic) [inline x/gamm/keeper/multihop_test.go:212]:
```suggestion
			// Prepare pools
```

File: x/gamm/keeper/multihop_test.go
Comment (czarcas7ic) [inline x/gamm/keeper/multihop_test.go:267]: Same comment on this as above

File: x/gamm/keeper/multihop_test.go
Comment (czarcas7ic) [inline x/gamm/keeper/multihop_test.go:286]: Same comment on this as above

File: CHANGELOG.md
Comment (czarcas7ic) [inline CHANGELOG.md:80]: Its now the multihop functions that are modified

**Review by czarcas7ic — COMMENTED** [review]
I think this is a much cleaner / better implementation, nicely done! I just have a few more questions/points. After these addressed and we get one more set of eyes I think this will be good to go

File: x/gamm/keeper/multihop_test.go
Comment (hleb-albau) [reply x/gamm/keeper/multihop_test.go:136]: removed

File: x/gamm/keeper/multihop_test.go
Comment (hleb-albau) [reply x/gamm/keeper/multihop_test.go:122]: remade it with new logic of changing pool swap fee in cached context
previously it uses two pool pairs(what is not oblivious, i agree here.

File: CHANGELOG.md
Comment (hleb-albau) [reply CHANGELOG.md:80]: ops

**Comment by hleb-albau** [comment]
@czarcas7ic rework test one more time again

tried different versions with `diffA` `diffB`, seems that variants with comments the most readable. 
Speaking about spot price asserts - that test was here before. I don't know how to make it better, without repeating whole AMM math in test or hard-coding final values. To ensure test is safe, i added assert that multihop output the same as separated swaps. Maybe it is enough(or even further we can remove spot price asserts)?

File: x/gamm/keeper/multihop_test.go
Comment (czarcas7ic) [inline x/gamm/keeper/multihop_test.go:311]: I don't think we need to make this function public right?

**Review by czarcas7ic — APPROVED** [review]
I think this is in a good place now, well done! Approving and can merge once we get another set of eyes on this.

File: x/gamm/types/route.go
Comment (nicolaslara) [inline x/gamm/types/route.go:32]: Should this support routes longer than 2? We could check that every odd route passes `r.TokenInDenom == appparams.BaseCoinUnit`

Comment (hleb-albau) [reply x/gamm/types/route.go:32]: n general the idea that osmo will have a lot of pools X-OSMO
Then to Swap X->Y all swaps will looks like X -> OSMO -> Y
So to detect such swaps we take the first hop and check that it has output OSMO(that mean that input for the second hop will be osmo too)

    What if route has length of three which all includes Osmo in the routes

That possible only if we have second hop OSMO->OSMO, or i miss something?

File: x/gamm/keeper/multihop_test.go
Comment (hleb-albau) [reply x/gamm/keeper/multihop_test.go]: fixed

File: x/gamm/types/route.go
Comment (nicolaslara) [reply x/gamm/types/route.go:32]: I was thinking of X -> OSMO -> Y -> OSMO -> Z.

Those can probably be simplified to less hops if there are x/osmo and z/osmo pools, but it isn't guaranteed that they exist.

I don't have a very strong opinion here, just thought it would make sense to future proof the iteration

Comment (sunnya97) [reply x/gamm/types/route.go:32]: `X -> OSMO -> Y -> OSMO -> Z` would imply that there are two `Y/OSMO` pools?  Which I think is a case we don't really have to worry about rn

Comment (hleb-albau) [reply x/gamm/types/route.go:32]: Also, knowing that there is such mechanic in place, tokens will be listed to OSMO and immediately will be available to all other assets for swaps with discount.

**Comment by hleb-albau** [comment]
@nicolaslara @sunnya97 @stackman27, so any new comments, or can we merge it?

**Review by stackman27 — APPROVED** [review]
@hleb-albau no more comments, I think everything looks solid!! 

Thank you for this feature :))

**Comment by nicolaslara** [comment]
No further comments on my end. The PR looks good to me.

**Comment by hleb-albau** [comment]
So...

**Review by mattverse — APPROVED** [review]

**Review by nicolaslara — APPROVED** [review]

File: x/gamm/keeper/multihop.go
Comment (hleb-albau) [reply x/gamm/keeper/multihop.go]: done

File: x/gamm/keeper/multihop_test.go
Comment (hleb-albau) [reply x/gamm/keeper/multihop_test.go:73]: Sorry, I don't understand what you suggest here :(

**Comment by hleb-albau** [comment]
rebased for new v12 main

**Comment by czarcas7ic** [comment]
Merging this! This however is not going to make it to v12, simply to reduce the sheer number of changes that need to get reviewed / tested before release.

**Comment by hleb-albau** [comment]
> Merging this! This however is not going to make it to v12, simply to reduce the sheer number of changes that need to get reviewed / tested before release.

It is so small, plus 100% covered by unit tests, plus viewed by at least 6 eyes. Also, manual test here is simple and requires to check only a few cases. Maybe it could be included?

**Comment by sunnya97** [comment]
We need to add support for this in the frontend as well

**Comment by jonator** [comment]
@sunnya97 https://github.com/osmosis-labs/osmosis-frontend/issues/745

**Pull request merged**
